Use cascade-summation for floats to avoid numerical instability

peterbell10 opened the pull request:
Fixes #38716, fixes #37234

This uses a variant of pairwise-summation to calculate sums of long lists without noticable loss of precison. It uses with multiple "levels" of accumulator along each axis, each of which is designed to hold the sum of an order of magnitude more values than the previous. 

e.g. if there are 2^16 elements, the first level will hold the sum of 2^4 elements, and so on in increasing powers of 2: 2^4, 2^8, 2^12 and finally 2^16.

This limits the differences in magnitude of the partial results being added together, and so we don't lose accuracy as the axis length increases.

Ref: https://en.wikipedia.org/wiki/Pairwise_summation

peterbell10 committed:
Add tree-based sum to avoid numerical instability

peterbell10 committed:
Fix second axis reductions

peterbell10 committed:
Support non-reduction loops inside parallel_reduce

peterbell10 committed:
Add vectorization to sum kernel

peterbell10 committed:
Better utilize last highest level accumulator

peterbell10 committed:
Add sum benchmark

peterbell10 committed:
vectorized_outer_sum now calculates multiple row sums in parallel

This is a direct analog of reduce128, but with the tree accumulator

peterbell10 commented:
I've updated the `vectorized_outer_sum` to more closely match `reduce128`. The updated benchmark figures now show at worst a 10% speed regression and still show up to 6x speed improvements for some non-contiguous inputs.

<details>
<summary>Full benchmark numbers below fold</summary>

```
# Input: R: 1024, V: 512, dim: 0, contiguous: True, device: cpu
This PR Execution Time (us) : 47.166
PT 1.5.0 Execution Time (us) : 46.152

# Input: R: 1024, V: 512, dim: 0, contiguous: False, device: cpu
This PR Execution Time (us) : 2930.226
PT 1.5.0 Execution Time (us) : 2658.398

# Input: R: 1024, V: 512, dim: 1, contiguous: True, device: cpu
This PR Execution Time (us) : 45.676
PT 1.5.0 Execution Time (us) : 44.131

# Input: R: 1024, V: 512, dim: 1, contiguous: False, device: cpu
This PR Execution Time (us) : 195.049
PT 1.5.0 Execution Time (us) : 1081.341

# Input: R: 1024, V: 1024, dim: 0, contiguous: True, device: cpu
This PR Execution Time (us) : 95.332
PT 1.5.0 Execution Time (us) : 93.709

# Input: R: 1024, V: 1024, dim: 0, contiguous: False, device: cpu
This PR Execution Time (us) : 6161.287
PT 1.5.0 Execution Time (us) : 5741.509

# Input: R: 1024, V: 1024, dim: 1, contiguous: True, device: cpu
This PR Execution Time (us) : 87.839
PT 1.5.0 Execution Time (us) : 84.595

# Input: R: 1024, V: 1024, dim: 1, contiguous: False, device: cpu
This PR Execution Time (us) : 535.933
PT 1.5.0 Execution Time (us) : 2249.186

# Input: R: 8192, V: 512, dim: 0, contiguous: True, device: cpu
This PR Execution Time (us) : 1435.037
PT 1.5.0 Execution Time (us) : 1460.735

# Input: R: 8192, V: 512, dim: 0, contiguous: False, device: cpu
This PR Execution Time (us) : 36749.245
PT 1.5.0 Execution Time (us) : 36095.155

# Input: R: 8192, V: 512, dim: 1, contiguous: True, device: cpu
This PR Execution Time (us) : 1086.657
PT 1.5.0 Execution Time (us) : 1029.092

# Input: R: 8192, V: 512, dim: 1, contiguous: False, device: cpu
This PR Execution Time (us) : 2194.304
PT 1.5.0 Execution Time (us) : 8854.416

# Input: R: 8192, V: 1024, dim: 0, contiguous: True, device: cpu
This PR Execution Time (us) : 4121.247
PT 1.5.0 Execution Time (us) : 4166.985

# Input: R: 8192, V: 1024, dim: 0, contiguous: False, device: cpu
This PR Execution Time (us) : 78065.462
PT 1.5.0 Execution Time (us) : 78299.460

# Input: R: 8192, V: 1024, dim: 1, contiguous: True, device: cpu
This PR Execution Time (us) : 2154.252
PT 1.5.0 Execution Time (us) : 2145.106

# Input: R: 8192, V: 1024, dim: 1, contiguous: False, device: cpu
This PR Execution Time (us) : 4428.362
PT 1.5.0 Execution Time (us) : 17697.258
```

</details>

A minor refactoring also revealed a flaw in how I was using the first level accumulator. Fixing this gave a further huge improvement in accuracy of large arrays:

![Errors](https://user-images.githubusercontent.com/13238737/84391638-2980db00-abf1-11ea-85b5-481a59d146e9.png)

peterbell10 committed:
Improve cache locality in scalar summation

peterbell10 commented:
From the latest commit, non-vectorized cases are 3-6x faster than pytorch 1.5. The vectorised cases are all measured the same, to within a 1-2% margin of error.

<details>
<summary>Graphs</summary>

![Quick](https://user-images.githubusercontent.com/13238737/84539986-67aef500-acec-11ea-932b-cd35d52b7da4.png)
![Medium](https://user-images.githubusercontent.com/13238737/84539993-6bdb1280-acec-11ea-906d-20db70c681e1.png)
![Slow](https://user-images.githubusercontent.com/13238737/84539997-6da4d600-acec-11ea-8ad8-a548c0ead305.png)

</details>

ngimel commented:
That's awesome! Do you think more comprehensive correctness tests are needed, especially for reductions over tuples of dimensions (they can't always be coalesced into reduction over a single dimension)? `sum` is one of the most common operations, and I'm not at all sure that our test suite covers all there is to cover.

peterbell10 commented:
It looks like tuples of dimensions are tested in `test_sum_dim` and also in `test_sum_noncontig`. Those test 1, 2, 3 dimension and all-dim reductions for contiguous and non-contiguous dimensions as well as contiguous and non-contiguous tensors. So, I think there is actually good coverage here already.

peterbell10 committed:
Avoid floating point logarithm in level step calculation

peterbell10 committed:
Add tolerance to test clip value

The two methods calculate the norm via different routes, so can be expected to
differ by more than a few ULP.

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+  static Vec256<scalar_t> load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto *ptr = data + index * stride;
+    return Vec256<scalar_t>::loadu(ptr);
+  }
```
It's a little weird that these implementations don't already exist in the vec256 headers. Would they logically belong there? Why not?

peterbell10 replied:
I couldn't see anything similar. It can't find any other kernels that can be templated on scalar or vector types and instead most have specialised code paths for each case.

ezyang commented on test/cpp/api/nn_utils.cpp:
```diff
@@ -67,7 +67,7 @@ TEST_F(NNUtilsTest, ClipGradNorm) {
     auto norm_after = compute_norm(norm_type);
     ASSERT_FLOAT_EQ(norm, norm_before);
-    ASSERT_FLOAT_EQ(norm_after, max_norm);
+    ASSERT_NEAR(norm_after, max_norm, 1e-6);
```
What's going on here?

peterbell10 replied:
This file cacluates the norm as `pow(p).sum()` whereas `clip_grad_norm` uses `norm(p).pow(p)`. Since `norm` uses a basic running sum, it rounds slightly differently.

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+    const int64_t row_stride,
+    const int64_t col_stride,
+    const int64_t size) {
+  constexpr int64_t num_levels = 4;
```
This is a tunable parameter, right?

peterbell10 replied:
Yes, adding more accumulator levels improves accuracy for larger arrays but may increase register pressure.

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+  return static_cast<int64_t>(llvm::findLastSet(ux - 1)) + 1;
+}
+
+// Simultaneously sum over n rows at once
```
I personally find it helpful when there's pseudocode for the "naive" version of this algorithm around, it helps me understand how the input values are put together. In this case it should be pretty simple, right?

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+    return;
+  }
+
+  // Custom floating point sum for better accuracy
```
There should be a high level description of what the algorithm does, similar to what you've already in the PR description, in the source code here, to make it easier for people to figure out what is going on.

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+    partial_sums[0] += load<scalar_t>(in_data, in_stride, i);
+  }
+
+  for (int64_t k = 1; k < ilp_factor; ++k) {
```
OOC: how come you don't need an unroll here?

peterbell10 replied:
I have mainly unrolled operations I expect to execute in parallel on the CPU hardware. Since there is a data dependency between each trip of the loop, I don't expect this to pipeline well in hardware.

ezyang commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+template <typename scalar_t>
+scalar_t row_sum(const char * C10_RESTRICT in_data,
+                 const int64_t in_stride, const int64_t size) {
+  constexpr int64_t ilp_factor = 4;
```
I notice `row_sum` is called with both a non-vectorized scalar type, and a vectorized scalar type. Does the ILP factor matter when you're not vectorizing? Why or why not?

peterbell10 replied:
To the best of my knowledge, vectorised and scalar additions have roughly the same characteristics in terms of register pressure and instruction latency.  Both can be pipelined in the CPU hardware to achieve greater parallelism. I expect this is the main reason why my implementation is so much faster in non-vectorised cases, since there was no ILP before.

ezyang commented:
I'm not qualified to review the reduction algorithm here. @colesbury / @ngimel do you think either of you would be able to do a detailed review? The new algorithm in SumKernel.cpp seems to be quite involved.

@peterbell10 it might be easier to review if we can get a more detailed algorithmic description, expanding beyond the PR description. Some things I'd be interested in seeing:

1. How closely related is your new `sum_kernel_impl` to the old `binary_kernel_reduce_vec`? I see a number of parallels, such as the treatment of inner/outer reductions separately. Would it be accurate to say we've preserved the high level structure of `binary_kernel_reduce_vec` entirely, but replaced `binary_kernel_reduce_vec`?
2. You seem to reuse code between the vectorized and non-vectorized implementations. How does this work? Do we get the expected behavior in the non-vectorized case?
3. A diagram showing how the inner / outer reductions operate on the data (with rows involved) I think would be really useful for someone trying to understand how the algorithm works in general.

peterbell10 committed:
Add comment explaining sum algorithm with simplified implementation

peterbell10 commented:
I've added a more detailed description of the sum algorithm in a comment on `multi_row_sum`. I will work on a diagram of the multidimensional sum structure  but it is more or less the same as the vectorised paths in `binary_kernel_reduce_vec`.

In fact, I could probably reuse a lot more of the existing reduction machinery if it was templated on a `multi_row_reduce` operation instead of binary ops. I could work on that change here, or in another PR if we want to remove the duplication.

ngimel commented:
I think it makes sense to remove duplication, but you can work on it in another PR. It would also be great to switch operations that currently use non-vectorized reduction kernel (norm, std etc) to your implementation, because non-vectorized reduction performance is pretty bad.

peterbell10 commented:
I have made this diagram, hopefully this makes it clearer:



![mulri_row_reduction](https://user-images.githubusercontent.com/13238737/84683365-03c83e80-af2f-11ea-82cf-d7d80eaf0319.png)

peterbell10 commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+          int64_t out_strides[] = { strides[0], strides[2] };
+
+          // Move reduction to be the 1st dim
+          if (out_strides[0] != 0 && out_strides[1] == 0) {
```
This doesn't seem to always happen. For example, 
```python
torch.rand(100, 1).sum(1)
```
Gives me strides `4, 0`.

ngimel replied:
hmhm, that's probably because when you are reducing over a dimension of size 1, is_dim_reduced does not identify it as reduction dimension.
https://github.com/pytorch/pytorch/blob/4c7d81f8479bce320cc11d1eb3adaf8ab0b90099/aten/src/ATen/native/TensorIterator.cpp#L472-L479
TensorIterator logic around reductions is pretty confusing :-(

peterbell10 replied:
For the `basic_loop` version of the reduction, that would give a better memory access pattern over the input vector. I'm guess that's why it's done this way?

ngimel replied:
I've dug a bit deeper, and it turns out in this case dimensions are coalesced in TI, so instead of (100,1) tensor you have (100) tensor. Then bogus (0,0) strides are pushed back to the end of the strides vector to fill in, and a bogus size1=1 is sent to the loop. Depending on whether input tensor is contiguous or not, vectorized_outer_sum or scalar_inner_sum will be called, both will produce correct results. Still don't know how out_strides can both be non-zero, but maybe it happens when `out` kwarg is specified and for some reason strides for dimension equal to 1 are not reset (their values don't matter).

peterbell10 replied:
I found a minimal reproducer for the non-zero strided reduction. It also relies on reduction over a length-1 dimension:
```python
torch.rand(2).expand(1, 3, 2).sum(0)
```

peterbell10 commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,304 @@
+          }
+
+          // Special case? - not a true reduction
+          if (out_strides[0] != 0 && out_strides[1] != 0) {
```
I'm not entirely sure but it comes up in one of the `test_autograd.py` test cases.

peterbell10 committed:
Rename store_result -> accumulate_result

ngimel commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -0,0 +1,336 @@
+      }
+
+      const auto mask = (level_mask << (j * level_power));
+      if ((i & mask) != mask) {
```
I think this should be 
```
      if ((i & mask) != 0) {
        break;
      }
```
otherwise the first failover (going to j=2) happens when i=2**(level_power-1)*level_step instead of 2**level_power*level_step. It does not matter in any practical terms, and code is functionally correct as written, it's just slightly confusing.

peterbell10 committed:
Fix off by one error

ngimel commented:
I think this should be good to go, although replacing the most common reduction is scary :-)

peterbell10 commented:
I've learned this algorithm has an established name, [pairwise-summation or cascade-summation](https://en.wikipedia.org/wiki/Pairwise_summation) which lead to numpy/numpy#3685. So, it seems NumPy actually uses this exact algorithm already with slightly different parameters.

ngimel commented:
@peterbell10 FYI, one of captum tests is failing with this PR, to repro, after installing captum go to tests/attr folder and run 
```
pytest -v -s test_feature_permutation.py
```
In all likelihood, the problem is not with this PR, and rather with captum tests relying on fp arithmetics that cannot be relied on, but I can't land with a failing test. I'm looking at the test in more detail.
